taskcluster/client.py

```diff
@@ -219,62 +222,125 @@ def genBewit():
             raise exceptions.TaskclusterFailure('Did not receive a bewit')
 
         u = urllib.parse.urlparse(requestUrl)
+
+        qs = u.query
+        if qs:
+            qs += '&'
+        qs += 'bewit=%s' % bewit
+
         return urllib.parse.urlunparse((
             u.scheme,
             u.netloc,
             u.path,
             u.params,
-            u.query + 'bewit=%s' % bewit,
+            qs,
             u.fragment,
         ))
 
     def _makeApiCall(self, entry, *args, **kwargs):
         """ This function is used to dispatch calls to other functions
         for a given API Reference entry"""
 
-        payload = None
-        _args = list(args)
-        _kwargs = copy.deepcopy(kwargs)
-
-        if 'input' in entry:
-            if len(args) > 0:
-                payload = _args.pop()
-            else:
-                raise exceptions.TaskclusterFailure('Payload is required as last positional arg')
-        apiArgs = self._processArgs(entry, *_args, **_kwargs)
-        route = self._subArgsInRoute(entry, apiArgs)
+        routeParams, payload, query = self._processArgs(entry, *args, **kwargs)
+        route = self._subArgsInRoute(entry, routeParams)
+        if query:
+            route += '?' + urllib.parse.urlencode(query)
         log.debug('Route is: %s', route)
 
         return self._makeHttpRequest(entry['method'], route, payload)
 
-    def _processArgs(self, entry, *args, **kwargs):
-        """ Take the list of required arguments, positional arguments
-        and keyword arguments and return a dictionary which maps the
-        value of the given arguments to the required parameters.
-
-        Keyword arguments will overwrite positional arguments.
+    def _processArgs(self, entry, *_args, **_kwargs):
+        """ Figure out, given an entry, positional and keyword arguments, what
+        the query-string options, payload and api arguments are.
         """
 
+        # We need the args to be a list so we can mutate them
+        args = list(_args)
+        kwargs = copy.deepcopy(_kwargs)
+        log.debug(args)
+        log.debug(kwargs)
+
         reqArgs = entry['args']
-        data = {}
+        routeParams = {}
+
+        query = {}
+        payload = None
+        kwApiArgs = {}
+
+        # There are three formats for calling methods:
+        #   1. method(v1, v1, payload)
```

Comment:
Do this mode not support query params?

Note: I ask because this is the mode we use in all other client libraries.
Typically, it looks like: `service.method(v1, v2, payload, query)` where `query` is optional, and `payload` is required if a payload is accepted by the method.

Reply:
it doesn't, and it cannot.  In other languages that's simple because you know how many position arguments are required, and for sure if payload is expected, so you can make the last arg optional.  In the python client, this is complicated by supportingn bare kwargs and args in the same call.

I think it's even better to force `query={}` because it just makes it so much clearer what's going on.